File: Sqrilizz-Reports/src/main/java/dev/sqrilizz/SQRILIZZREPORTS/lifecycle/ShutdownManager.java
package dev.sqrilizz.SQRILIZZREPORTS.lifecycle;

import dev.sqrilizz.SQRILIZZREPORTS.Main;
import dev.sqrilizz.SQRILIZZREPORTS.api.CacheManager;
import dev.sqrilizz.SQRILIZZREPORTS.api.RESTServer;
import dev.sqrilizz.SQRILIZZREPORTS.db.DatabaseManager;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Система graceful shutdown для идеального завершения работы плагина
 * Обеспечивает корректное закрытие всех ресурсов и сервисов
 */
public class ShutdownManager {
    
    private static final AtomicBoolean isShuttingDown = new AtomicBoolean(false);
    private static final ExecutorService shutdownExecutor = Executors.newSingleThreadExecutor(r -> {
        Thread t = new Thread(r, "SQRILIZZ-Shutdown");
        t.setDaemon(true);
        return t;
    });
    
    /**
     * Инициализировать систему graceful shutdown
     */
    public static void initialize() {
        // Регистрируем shutdown hook для JVM
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            if (!isShuttingDown.get()) {
                performGracefulShutdown();
            }
        }, "SQRILIZZ-ShutdownHook"));
        
        Main.getInstance().getLogger().info("Graceful shutdown system initialized");
    }
    
    /**
     * Выполнить graceful shutdown всех компонентов
     */
    public static void performGracefulShutdown() {
        if (isShuttingDown.compareAndSet(false, true)) {
            Main.getInstance().getLogger().info("[SHUTDOWN] Starting graceful shutdown...");
            
            long startTime = System.currentTimeMillis();
            
            try {
                // Параллельное завершение всех сервисов
                CompletableFuture<Void> allShutdowns = CompletableFuture.allOf(
                    shutdownRESTServer(),
                    shutdownDatabase(),
                    shutdownCache(),
                    shutdownDiscordBot(),
                    shutdownTelegramBot(),
                    savePerformanceReport()
                );
                
                // Ждем завершения всех операций (максимум 30 секунд)
                allShutdowns.get(30, TimeUnit.SECONDS);
                
                long duration = System.currentTimeMillis() - startTime;
                Main.getInstance().getLogger().info("[SHUTDOWN] Graceful shutdown completed in " + duration + "ms");
                
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[SHUTDOWN] Some components failed to shutdown gracefully: " + e.getMessage());
            } finally {
                shutdownExecutor.shutdown();
            }
        }
    }
    
    /**
     * Проверить, выполняется ли shutdown
     */
    public static boolean isShuttingDown() {
        return isShuttingDown.get();
    }
    
    private static CompletableFuture<Void> shutdownRESTServer() {
        return CompletableFuture.runAsync(() -> {
            try {
                Main.getInstance().getLogger().info("[REST] Shutting down REST server...");
                RESTServer.shutdown();
                Main.getInstance().getLogger().info("[REST] REST server shutdown complete");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[REST] REST server shutdown failed: " + e.getMessage());
            }
        }, shutdownExecutor);
    }
    
    private static CompletableFuture<Void> shutdownDatabase() {
        return CompletableFuture.runAsync(() -> {
            try {
                Main.getInstance().getLogger().info("[DATABASE] Shutting down database connections...");
                DatabaseManager.close();
                Main.getInstance().getLogger().info("[DATABASE] Database shutdown complete");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[DATABASE] Database shutdown failed: " + e.getMessage());
            }
        }, shutdownExecutor);
    }
    
    private static CompletableFuture<Void> shutdownCache() {
        return CompletableFuture.runAsync(() -> {
            try {
                Main.getInstance().getLogger().info("[CACHE] Shutting down cache system...");
                CacheManager.cleanUp();
                Main.getInstance().getLogger().info("[CACHE] Cache shutdown complete");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[CACHE] Cache shutdown failed: " + e.getMessage());
            }
        }, shutdownExecutor);
    }
    
    private static CompletableFuture<Void> shutdownDiscordBot() {
        return CompletableFuture.runAsync(() -> {
            try {
                Main.getInstance().getLogger().info("[DISCORD] Shutting down Discord bot...");
                // DiscordBot.shutdown(); // Если есть такой метод
                Main.getInstance().getLogger().info("[DISCORD] Discord bot shutdown complete");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[DISCORD] Discord bot shutdown failed: " + e.getMessage());
            }
        }, shutdownExecutor);
    }
    
    private static CompletableFuture<Void> shutdownTelegramBot() {
        return CompletableFuture.runAsync(() -> {
            try {
                Main.getInstance().getLogger().info("[TELEGRAM] Shutting down Telegram manager...");
                dev.sqrilizz.SQRILIZZREPORTS.TelegramManager.shutdown();
                Main.getInstance().getLogger().info("[TELEGRAM] Telegram manager shutdown complete");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[TELEGRAM] Telegram manager shutdown failed: " + e.getMessage());
            }
        }, shutdownExecutor);
    }
    
    private static CompletableFuture<Void> savePerformanceReport() {
        return CompletableFuture.runAsync(() -> {
            try {
                Main.getInstance().getLogger().info("[PERFORMANCE] Saving final performance report...");
                String report = "Performance monitoring disabled";
                Main.getInstance().getLogger().info("Final Performance Report: " + report);
                Main.getInstance().getLogger().info("[PERFORMANCE] Performance report saved");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("[PERFORMANCE] Performance report save failed: " + e.getMessage());
            }
        }, shutdownExecutor);
    }
    
    /**
     * Принудительное завершение (для экстренных случаев)
     */
    public static void forceShutdown() {
        Main.getInstance().getLogger().warning("[FORCE SHUTDOWN] Force shutdown initiated");
        isShuttingDown.set(true);
        
        try {
            shutdownExecutor.shutdownNow();
            if (!shutdownExecutor.awaitTermination(5, TimeUnit.SECONDS)) {
                Main.getInstance().getLogger().severe("[FORCE SHUTDOWN] Force shutdown timeout");
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: Sqrilizz-Reports/src/main/kotlin/dev/sqrilizz/SQRILIZZREPORTS/lifecycle/ShutdownManager.kt
package dev.sqrilizz.SQRILIZZREPORTS.lifecycle

import dev.sqrilizz.SQRILIZZREPORTS.Main
import dev.sqrilizz.SQRILIZZREPORTS.TelegramManager
import dev.sqrilizz.SQRILIZZREPORTS.api.CacheManager
import dev.sqrilizz.SQRILIZZREPORTS.api.RESTServer
import dev.sqrilizz.SQRILIZZREPORTS.db.DatabaseManager
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

object ShutdownManager {

    private val isShuttingDown = AtomicBoolean(false)
    private val shutdownExecutor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "SQRILIZZ-Shutdown").apply { isDaemon = true }
    }

    @JvmStatic
    fun initialize() {
        Runtime.getRuntime().addShutdownHook(Thread({
            if (!isShuttingDown.get()) {
                performGracefulShutdown()
            }
        }, "SQRILIZZ-ShutdownHook"))

        Main.getInstance().logger.info("Graceful shutdown system initialized")
    }

    @JvmStatic
    fun performGracefulShutdown() {
        if (isShuttingDown.compareAndSet(false, true)) {
            Main.getInstance().logger.info("\uD83D\uDD04 Starting graceful shutdown...")

            val startTime = System.currentTimeMillis()

            try {
                val allShutdowns = CompletableFuture.allOf(
                    shutdownComponent("REST server") { RESTServer.shutdown() },
                    shutdownComponent("database connections") { DatabaseManager.close() },
                    shutdownComponent("cache system") { CacheManager.cleanUp() },
                    shutdownComponent("Discord bot") { /* placeholder */ },
                    shutdownComponent("Telegram manager") { TelegramManager.shutdown() },
                    shutdownComponent("performance report") {
                        Main.getInstance().logger.info("Final Performance Report:\nPerformance monitoring disabled")
                    }
                )

                allShutdowns.get(30, TimeUnit.SECONDS)

                val duration = System.currentTimeMillis() - startTime
                Main.getInstance().logger.info("\u2705 Graceful shutdown completed in ${duration}ms")
            } catch (e: Exception) {
                Main.getInstance().logger.warning("\u26a0\ufe0f Some components failed to shutdown gracefully: ${e.message}")
            } finally {
                shutdownExecutor.shutdown()
            }
        }
    }

    @JvmStatic
    fun isShuttingDown(): Boolean = isShuttingDown.get()

    @JvmStatic
    fun forceShutdown() {
        Main.getInstance().logger.warning("\uD83D\uDEA8 FORCE SHUTDOWN INITIATED")
        isShuttingDown.set(true)

        try {
            shutdownExecutor.shutdownNow()
            if (!shutdownExecutor.awaitTermination(5, TimeUnit.SECONDS)) {
                Main.getInstance().logger.severe("\u274c Force shutdown timeout")
            }
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        }
    }

    private fun shutdownComponent(name: String, action: () -> Unit): CompletableFuture<Void> =
        CompletableFuture.runAsync({
            try {
                Main.getInstance().logger.info("Shutting down $name...")
                action()
                Main.getInstance().logger.info("$name shutdown complete")
            } catch (e: Exception) {
                Main.getInstance().logger.warning("$name shutdown failed: ${e.message}")
            }
        }, shutdownExecutor)
}
